src/vs/platform/contextkey/common/contextkey.ts

```diff
@@ -241,6 +269,370 @@ export abstract class ContextKeyExpr {
+							}
+						}
+					}
+
+					return ContextKeyRegexExpr.create(key, regex);
+				} else {
+					throw this._errExpectedButGot('REGEX', this._peek());
+				}
+			}
+
+			// [ 'not' 'in' value ]
+			if (this._match(TokenType.Not)) {
+				this._consume(TokenType.In, `'in' after 'not'`);
+				const right = this._value();
+				return ContextKeyExpr.notIn(key, right);
+			}
+
+			// [ ('==' | '!=' | '<' | '<=' | '>' | '>=' | 'in') value ]
+			if (this._match(TokenType.Eq, TokenType.NotEq, TokenType.Lt, TokenType.LtEq, TokenType.Gt, TokenType.GtEq, TokenType.In)) {
+				const op = this._previous().type;
+				const right = this._value();
+				switch (op) {
+					case TokenType.Eq: {
+						switch (right) {
+							case 'true':
+								return ContextKeyExpr.has(key);
+							case 'false':
+								return ContextKeyExpr.not(key);
+							default:
+								return ContextKeyExpr.equals(key, right);
+						}
+					}
+					case TokenType.NotEq: {
+						switch (right) {
+							case 'true':
+								return ContextKeyExpr.not(key);
+							case 'false':
+								return ContextKeyExpr.has(key);
+							default:
+								return ContextKeyExpr.notEquals(key, right);
+						}
+					}
+
+					// TODO: ContextKeyExpr.smaller(key, right) accepts only `number` as `right` AND during eval of this node, we just eval to `false` if `right` is not a number
+					// consequently, package.json linter should _warn_ the user if they're passing undesired things to ops
+					case TokenType.Lt: return ContextKeySmallerExpr.create(key, right);
+					case TokenType.LtEq: return ContextKeySmallerEqualsExpr.create(key, right);
+					case TokenType.Gt: return ContextKeyGreaterExpr.create(key, right);
+					case TokenType.GtEq: return ContextKeyGreaterEqualsExpr.create(key, right);
+
+					case TokenType.In: return ContextKeyExpr.in(key, right);
+
+					default:
+						throw illegalState(`must've matched the op with this._match() call above`);
+				}
+			}
+
+			return ContextKeyExpr.has(key);
+		} else {
+			throw this._errExpectedButGot(`'true', 'false', '(', KEY, KEY '=~' regex, KEY [ ('==' | '!=' | '<' | '<=' | '>' | '>=' | 'in' | 'not' 'in') value ]`, this._peek());
+		}
+	}
+
+	private _value(): string { // TODO@ulugbekna: match all at once and then `switch` on the type
+		if (this._match(TokenType.Str, TokenType.QuotedStr)) {
+			return this._previous().lexeme!;
+		} if (this._match(TokenType.True)) {
+			return 'true';
+		} if (this._match(TokenType.False)) {
+			return 'false';
+		} if (this._match(TokenType.In)) { // we support `in` as a value, e.g., "when": "languageId == in" - exists in existing extensions
+			return 'in';
+		} else {
+			return ''; // this allows "when": "foo == " which's used by existing extensions
+		}
+	}
+
+	// careful: this can throw if current token is the initial one (ie index = 0)
+	private _previous() {
+		return this._tokens[this._current - 1];
+	}
+
+	private _match(...tokens: TokenType[]) {
```

Comment:
AFAIK functions with for loops rarely get inlined. I'd like to suggest to create two versions, `_matchOne` and `_matchAny` and use each on of them as needed directly. I think this is a low hanging perf improvement.